File: start-accessibility/dotnet/Helpers/ResponseStreamer.cs
using GitHub.Copilot;

namespace HelloCopilotSDK.Helpers;

public static class ResponseStreamer
{
    public static async Task SendAndPrintAsync(
        CopilotSession session,
        string prompt,
        TimeSpan? timeout = null,
        CancellationToken cancellationToken = default)
    {
        var actualTimeout = timeout ?? TimeSpan.FromSeconds(120);
        if (actualTimeout <= TimeSpan.Zero)
        {
            throw new ArgumentOutOfRangeException(nameof(timeout), "The response timeout must be positive.");
        }
        using var deadline = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        deadline.CancelAfter(actualTimeout);
        var completed = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
        var receivedDelta = false;

        using var subscription = session.On<SessionEvent>(sessionEvent =>
        {
            if (completed.Task.IsCompleted)
            {
                return;
            }
            switch (sessionEvent)
            {
                case AssistantMessageDeltaEvent delta when !string.IsNullOrEmpty(delta.Data.DeltaContent):
                    receivedDelta = true;
                    Console.Write(delta.Data.DeltaContent);
                    break;
                case AssistantMessageEvent message:
                    if (!receivedDelta)
                    {
                        Console.Write(message.Data.Content);
                    }
                    receivedDelta = false;
                    break;
                case ToolExecutionStartEvent tool:
                    Console.WriteLine($"\n[tool:start] {tool.Data.ToolName}");
                    break;
                case ToolExecutionCompleteEvent tool:
                    Console.WriteLine($"[tool:done] success={tool.Data.Success}");
                    break;
                case SessionIdleEvent:
                    Console.WriteLine();
                    completed.TrySetResult();
                    break;
                case SessionErrorEvent error:
                    completed.TrySetException(new InvalidOperationException(error.Data.Message));
                    break;
            }
        });

        try
        {
            var send = session.SendAsync(new MessageOptions { Prompt = prompt }, deadline.Token);
            var first = await Task.WhenAny(send, completed.Task).WaitAsync(deadline.Token);
            await first;
            await send.WaitAsync(deadline.Token);
            await completed.Task.WaitAsync(deadline.Token);
        }
        catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
        {
            throw new TimeoutException("The response reached the timeout.");
        }
        finally
        {
            await deadline.CancelAsync();
        }
    }
}

File: finished/java/accessibility-report/src/main/java/workshop/ResponseStreamer.java
package workshop;

import com.github.copilot.CopilotSession;
import com.github.copilot.generated.AssistantMessageDeltaEvent;
import com.github.copilot.generated.AssistantMessageEvent;
import com.github.copilot.generated.ToolExecutionStartEvent;
import com.github.copilot.generated.ToolExecutionCompleteEvent;
import com.github.copilot.rpc.MessageOptions;

import java.time.Duration;
import java.util.concurrent.atomic.AtomicBoolean;

public final class ResponseStreamer {
    private ResponseStreamer() {
    }

    public static void sendAndPrint(CopilotSession session, String prompt) throws Exception {
        sendAndPrint(session, prompt, Duration.ofSeconds(120));
    }

    public static void sendAndPrint(CopilotSession session, String prompt, Duration timeout) throws Exception {
        if (timeout.isNegative() || timeout.toMillis() == 0) {
            throw new IllegalArgumentException("The response timeout must be positive.");
        }
        var receivedDelta = new AtomicBoolean(false);
        try (var subscription = session.on(event -> {
            if (event instanceof AssistantMessageDeltaEvent delta && delta.getData() != null) {
                String content = delta.getData().deltaContent();
                if (content != null && !content.isEmpty()) {
                    receivedDelta.set(true);
                    System.out.print(content);
                }
            } else if (event instanceof AssistantMessageEvent message && message.getData() != null) {
                if (!receivedDelta.getAndSet(false) && message.getData().content() != null) {
                    System.out.print(message.getData().content());
                }
            } else if (event instanceof ToolExecutionStartEvent tool && tool.getData() != null) {
                System.out.println("\n[tool:start] " + tool.getData().toolName());
            } else if (event instanceof ToolExecutionCompleteEvent tool && tool.getData() != null) {
                System.out.println("[tool:done] success=" + tool.getData().success());
            }
        })) {
            // The SDK wait handles session.error, idle, send failures and timeout, and releases its listener.
            var response = session.sendAndWait(new MessageOptions().setPrompt(prompt), timeout.toMillis());
            try {
                response.get();
            } finally {
                response.cancel(true);
            }
            System.out.println();
        }
    }
}
